src/main/java/org/chameleoncloud/api/admin/UserGroupRolesResource.java

```diff
@@ -0,0 +1,320 @@
+		    grantAdminPermissions(userGroup.getParent(), userGroup);
+		} else if (userGroupName.endsWith(UserGroupRoles.MANAGER_GROUP_SUFFIX)) {
+		    grantManagerPermissions(userGroup.getParent(), userGroup);
+		}
+	    }
+	    userGroupIds.add(userGroup.getId());
+	}
+
+	Set<UserGroupRoles> result = new HashSet<UserGroupRoles>();
+	for (GroupModel group : userGroups) {
+
+	    if (isRootGroup(group)) {
+
+		UserGroupRoles ugr = new UserGroupRoles();
+		ugr.setGroupId(group.getId());
+		ugr.setGroupName(group.getName());
+		SortedMap<Integer, String> policies = new TreeMap<Integer, String>();
+		Set<String> scopes = new HashSet<>();
+
+		for (String permissionName : groupPermission.getPermissions(group).keySet()) {
+		    Policy permission = getPolicyByPermissionName(permissionName, group);
+		    SortedMap<Integer, String> groupPolicies = getPolicies(permission, userGroupIds);
+		    if (!groupPolicies.isEmpty()) {
+			policies.putAll(groupPolicies);
+			scopes.addAll(getPermissionScopes(permission));
+		    }
+		}
+
+		ugr.setPolicy(policies.get(policies.firstKey()));
+		ugr.setScopes(scopes);
+		result.add(ugr);
+	    }
+	}
+
+	return Response.status(Response.Status.OK).entity(result).build();
+    }
+
+    @PUT
+    @Path("{user}")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    public Response setUserGroupRole(UserGroupRoles userGroupRoles, @PathParam("user") String userId) {
+	UserModel user = session.users().getUserById(userId, realm);
+	if (user == null) {
+	    return Response.status(Response.Status.NOT_FOUND).entity("user not found!").build();
+	}
+
+	String requestGroupId = userGroupRoles.getGroupId();
+	String requestGroupName = userGroupRoles.getGroupName();
+	String requestPolicy = userGroupRoles.getPolicy();
+	if (requestPolicy == null) {
+	    return Response.status(Response.Status.NOT_FOUND).entity("policy not found!").build();
+	}
+
+	GroupModel group = null;
+	if (requestGroupId != null) {
+	    group = session.realms().getGroupById(requestGroupId, realm);
+	} else {
+	    group = session.realms().getGroups(realm).stream().filter(g -> requestGroupName.equals(g.getName()))
+		    .findFirst().orElse(null);
+	}
+	if (group == null) {
+	    return Response.status(Response.Status.NOT_FOUND).entity("group not found!").build();
+	}
+
+	if (!isRootGroup(group)) {
+	    return Response.status(Response.Status.INTERNAL_SERVER_ERROR).entity("Not a root group!").build();
+	}
+
+	user.joinGroup(group);
+	grantMemberPermissions(group);
+
+	GroupModel adminSubGroup = group.getSubGroups().stream()
+		.filter(g -> g.getName().endsWith(UserGroupRoles.ADMIN_GROUP_SUFFIX)).findFirst().orElse(null);
+	GroupModel managerSubGroup = group.getSubGroups().stream()
+		.filter(g -> g.getName().endsWith(UserGroupRoles.MANAGER_GROUP_SUFFIX)).findFirst().orElse(null);
+	if (requestPolicy.equals(UserGroupRoles.ADMIN)) {
+	    user.joinGroup(adminSubGroup);
```

Comment:
What if adminSubGroup is null? I think this will throw.

Reply:
yes. it will throw.

Reply:
It was not obvious in my comment, but what I meant was: I don't think this should throw a NPE (null pointer exception). NPE means the code hasn't dealt with the case. I think it would be better to check for null and either (a) do nothing and proceed, but log that this happened, or (b) raise a RuntimeException and abort the request. Which do you think is best?